src/Optimization/optimization_problem_results.jl

```diff
@@ -0,0 +1,829 @@
+    end
+
+    @debug "Exported $filename"
+    return
+end
+
+"""
+Exports all results from the operations problem.
+"""
+function export_results(results::OptimizationProblemResults; kwargs...)
+    exports = OptimizationProblemResultsExport(
+        "Problem";
+        store_all_duals = true,
+        store_all_parameters = true,
+        store_all_variables = true,
+        store_all_aux_variables = true,
+    )
+    return export_results(results, exports; kwargs...)
+end
+
+function export_results(
+    results::OptimizationProblemResults,
+    exports::OptimizationProblemResultsExport;
+    file_type = CSV.File,
+)
+    file_type != CSV.File && error("only CSV.File is currently supported")
+    for (source, decider, label) in [
+        (results.variable_values, should_export_variable, "variables"),
+        (results.aux_variable_values, should_export_aux_variable, "aux_variables"),
+        (results.dual_values, should_export_dual, "duals"),
+        (results.parameter_values, should_export_parameter, "parameters"),
+        (results.expression_values, should_export_expression, "expressions"),
+    ]
+        export_path = mkpath(joinpath(get_output_dir(results), label))
+        for (key, df) in source
+            if decider(exports, key)
+                export_result(file_type, export_path, key, df)
+            end
+        end
+    end
+
+    if exports.optimizer_stats
+        export_result(
+            file_type,
+            joinpath(get_output_dir(results), "optimizer_stats.csv"),
+            results.optimizer_stats,
+        )
+    end
+
+    @info "Exported OptimizationProblemResults to $(get_output_dir(results))"
+end
+
+function _deserialize_key(
+    ::Type{<:OptimizationContainerKey},
+    results::OptimizationProblemResults,
+    name::AbstractString,
+)
+    return deserialize_key(results.optimization_container_metadata, name)
+end
+
+function _deserialize_key(
+    ::Type{T},
+    ::OptimizationProblemResults,
+    args...,
+) where {T <: OptimizationContainerKey}
+    return make_key(T, args...)
+end
+
+function _validate_keys(existing_keys, result_keys)
+    diff = setdiff(result_keys, existing_keys)
+    if !isempty(diff)
+        throw(IS.InvalidValue("These keys are not stored: $diff"))
+    end
+    return
+end
+
+read_optimizer_stats(res::OptimizationProblemResults) = res.optimizer_stats
+
+"""
+Set the system in the results instance.
+
+Throws InvalidValue if the source UUID is incorrect.
+"""
+function set_source_data!(
+    res::OptimizationProblemResults,
+    source::InfrastructureSystemsType,
+)
+    source_uuid = get_uuid(source)
+    if source_uuid != res.source_data_uuid
+        throw(
+            InvalidValue(
+                "System mismatch. $sys_uuid does not match the stored value of $(res.source_uuid)",
+            ),
+        )
+    end
+
+    res.source_data = source
+    return
+end
+
+const _PROBLEM_RESULTS_FILENAME = "problem_results.bin"
+
+# TODO test this in IS
```

Comment:
TODO?

Reply:
Yes, we are going to need to add more tests for this in a later PR inside IS. In the short term we are going to have to rely on PSI to test this.

Reply:
https://github.com/NREL-Sienna/InfrastructureSystems.jl/issues/351